keps/sig-apps/0012-20180522-deployment-inplace-update.md

```diff
@@ -0,0 +1,198 @@
+
+	// Rolling update config params. Present only if DeploymentStrategyType =
+	// RollingUpdate.
+	//---
+	// TODO: Update this to follow our convention for oneOf, whatever we decide it
+	// to be.
+	// +optional
+	RollingUpdate *RollingUpdateDeployment
+
++	// InPlace update config params. Present only if DeploymentStrategyType =
++	// InPlaceUpdate.
++	// +optional
++	InPlaceUpdate *InPlaceUpdateDeployment
+}
+
+const(
+	// Kill all existing pods before creating new ones.
+	RecreateDeploymentStrategyType DeploymentStrategyType = "Recreate"
+
+        // Replace the old RCs by new one using rolling update i.e gradually scale down the old RCs and scale up the new one.
+        RollingUpdateDeploymentStrategyType DeploymentStrategyType = "RollingUpdate"
+    
+        // Update containers inside the pods without tearing down the pods
++	InPlaceUpdateDeploymentStrategyType DeploymentStrategyType = "InPlaceUpdate"
+)
+
+// Spec to control the desired behavior of in-place update. This is gated by readiness.
++ type InPlaceUpdateDeployment struct {
++	 // The maximum number of pods that can be unavailable during the update.
++	 // Value can be an absolute number (ex: 5) or a percentage of total pods at the start of update (ex: 10%).
++	 // Absolute number is calculated from percentage by rounding down.
++	 MaxUnavailable intstr.IntOrString
++ }
+```
+
+
+### Deployment Controller
+
+#### Update
+Below is the proposed approach for in-place update.
+Wait until expected number of pods created, when deployment controller receives the in-place update request
+
+```
+    newReplicaSet created with 0 replica
+    ...
+
+    // Find all old pods and update the unhealthy(failed, pending, not ready) pods first with the new spec from new ReplicaSet and then the running pods,
+    // MaxUnavailable is the maximum number of pods that can be unavailable during the update.
+    podsToUpdate = getPodsToUpdate(oldPods, MaxUnavailable)
+    for pod range(podsToUpdate)
+        update(pod, newSpec)
+
+    // swap oldReplicaSet and newReplicaSet information such as spec, revision, annotation, timestamp etc.
+    
+    // add oldReplicaSet as an annotation of newReplicaSet which also gets persisted into etcd, in case the controller failed that may cause oldReplicaSet lost. 
+    addAnnotationToNewReplicaSet(oldReplicaSet)
+    
+    // replace old ReplcaSet with the new ReplicaSet info. Since there are two ReplicaSet exists in the system, 
+    // we need to make sure that functions such as FindNewReplicaSet retrieves the right ReplicaSet. 
+    // we may also need to swap the timestamp too, to make sure the current ReplicaSet has the latest timestamp. 
+    oldReplicaSet = newReplicaSet
+    
+    // retrieve back the annotation for oldReplicaSet and replace newReplicaSet with the oldReplicaSet info.
+    // The oldReplicaSet info serves as a historic ReplicaSet for the purpose of rollback.
+    newReplicaSet = getAnnotation(oldReplicaSet)
+        
```

Comment:
You also need to add the last step of doing the in-place update of the pods

Reply:
Yep.  I had that in the beginning. Now I moved that to the end, as I think we should update the ReplicaSet info first, in case there's pod failed, a new pod created should be a pod with the new spec instead of old.